net/xia/ppal_ether/main.c

```diff
@@ -168,6 +169,316 @@ static void local_free_ether(struct fib_xid_table *xtbl, struct fib_xid *fxid)
 	kfree(leid);
 }
 
+/* Main-FXID. */
+struct fib_xid_ether_main {
+	struct xip_dst_anchor   xem_anchor;
+	struct net_device       *host_interface;
+	struct list_head        interface_common_addr;
+	/* Hardware address of the neighbor. */
+	u8                      ha[ALIGN(MAX_ADDR_LEN, sizeof(long))];
+
+	/* WARNING: @xhm_common is of variable size, and
+	 * MUST be the last member of the struct.
+	 */
+	struct fib_xid          xem_common;
+};
+
+/* Main-FXID functions. */
+static inline struct fib_xid_ether_main *fxid_mether(struct fib_xid *fxid)
+{
+	return likely(fxid)
+		? container_of(fxid, struct fib_xid_ether_main, xem_common)
+		: NULL;
+}
+
+static inline u8 ctou8(char byte)
+{
+	if (byte >= '0' && byte <= '9')
+		return byte - '0';
+	else if (byte >= 'a' && byte <= 'f')
+		return byte - 'a' + 10;
+	else if (byte >= 'A' && byte <= 'F')
+		return byte - 'A' + 10;
+	return -EINVAL;
+}
+
+static inline int get_xid_info(char *strid, u8 *lladdr)
+{
+	u32 ifindex;
+	int itr = 0, itr2 = 0;
+
+	ifindex = 0;
+	while (itr < 8) {
+		u8 val = ctou8(strid[itr++]) & 0xF;
+
+		if (val < 0)
+			return val;
+		ifindex = (ifindex << 4) | val;
+	}
+
+	while (itr < 20) {
+		u8 val = ctou8(strid[itr++]) & 0xF;
+
+		if (val < 0)
+			return val;
+		lladdr[itr2] = val;
+
+		val = ctou8(strid[itr++]) & 0xF;
+		if (val < 0)
+			return val;
+		lladdr[itr2] = (lladdr[itr2] << 4) | val;
+		itr2++;
+	}
+	return ifindex;
+}
+
+/* Main-FXID + INTERFACE. */
+static void attach_neigh_to_interface(struct fib_xid_ether_main *mether)
+{
+	struct ether_interface *einterface;
+
+	einterface = ether_interface_get(mether->host_interface);
+
+	/*When using list_add_tail_rcu the caller must take whatever precautions
+	* are necessary (such as holding appropriate locks) to avoid racing
+	* with another list-mutation primitive, such as list_add_tail_rcu()
+	* or list_del_rcu(), running on this same list.
+	*/
+	spin_lock(&einterface->interface_lock);
+	list_add_tail_rcu(&einterface->list_interface_common_addr,
+			  &mether->interface_common_addr);
+	spin_unlock(&einterface->interface_lock);
+
+	ether_interface_put(einterface);
+}
+
+/* Main-FXID - INTERFACE.. */
+static void detach_neigh_to_interface(struct fib_xid_ether_main *mether)
+{
+	struct ether_interface *einterface;
+
+	einterface = ether_interface_get(mether->host_interface);
```

Comment:
But isn't it that ether_interface_put returning NULL is possible only after eint_destroy is called? If the reason behind this is some other application doing so, then it could also acquire the eth_ptr and delete the data pointed to by eth_ptr too, so it is a different issue from that angle and should be dealt with separately rather than just shifting the pointer to ether interface to mfxid.

Reply:
To better appreciate the issue, consider the following execution line: an interface is removed. First, the interface goes down and all XIDs associated with it are removed, but the XID entries are kept in memory for a while because of RCU. Second, the interface is finally removed. In this state, `ether_interface_get()` returns NULL, but the XID entries are still pointing to the list in Ethernet interface. Even worse, since none of the XID entries have increased the reference counter of the Ethernet Interface, it goes away. Now all removed XID entries have a dangling pointer.